### ISSUE TITLE
Invalid json body always return 500 

### ISSUE THREAD

**Issue by kingdun3284** [FoalTS/foal#457]
It executes json parse before any hooks. So I simply can't write a hook to check the body is a valid json or not before returning an 500 error.

**Comment by LoicPoullain** [comment]
Hi @kingdun3284 

Thanks for your interest!

In which situation do you want to send an invalid JSON and check it yourself?

**Comment by LoicPoullain** [comment]
Also, what is the traceback of the error?

**Comment by kingdun3284** [comment]
`<html><head><title>INTERNAL SERVER ERROR</title></head><body><h1>500 - INTERNAL SERVER ERROR</h1><pre>SyntaxError: Unexpected end of JSON input
</pre></body></html>`

Basically it's fine as the server doesn't shut down due to this error. However, I want to customize the error message to tell the client that their body format is incorrect instead of having 500 internal server error which isn't meaningful most of the time. However, this error is triggered before any hooks are executed.

**Comment by LoicPoullain** [comment]
Sorry for the late reply.

I see one hacky solution to achieve this by providing an express instance to the `createApp` function  in `src/index.ts`:

```typescript
// Before
const app = createApp(AppController, { /* ... */ });

// After
const expressApp = express();
expressApp.use(express.json());
expressApp.use((err, req, res, next) => {
  if (err.status === 400) {
    res.status(400).send({
      body: err.body,
      message: 'Invalid JSON',
    });
    return;
  }
  next(err);
});
const app = createApp(AppController, { /* ... */ }, expressApp);
```

The idea here is to manually handle the JSON parsing before executing the code of the framework. It's not really pretty but I guess it would solve your problem for now?

**Comment by yaaminu** [comment]
@LoicPoullain don't you think that this suggests that the best way to handle errors of these kinds is to write our own body parsing middleware? We could start with a well-known module like `body-parser`

**Comment by LoicPoullain** [comment]
`express.json()` uses `body-parser` under the hood. It is the only middleware that is included in ExpressJS by default.

FoalTS uses it internally but the error is not handled properly. It is a bug. It should be fixed in the next release. 👍

**Comment by yaaminu** [comment]
> FoalTS uses it internally but the error is not handled properly. It is a bug. It should be fixed in the next release. 

Nice. Have you made the fix already or you're yet to?

**Comment by LoicPoullain** [comment]
I'm already working on it. The issue covers a wider field than it seems at first glance. It points out that currently all middleware errors (400, 415 for charset, etc) return a 500 which is not the expected behavior. I need to look more closely at the content of middleware errors to make sure that the framework is not missing anything.

**Comment by LoicPoullain** [comment]
Fixed in v1.0.1 (published this week or the next one)
